importexport/musicxml/importmxmlpass2.cpp

```diff
@@ -910,42 +911,50 @@ static void handleTupletStop(Tuplet*& tuplet, const int normalNotes)
 //   addArticulationToChord
 //---------------------------------------------------------
 
-static void addArticulationToChord(ChordRest* cr, SymId articSym, QString dir)
+static void addArticulationToChord(ChordRest* cr, SymId articSym, QString dir, QString place)
       {
       Articulation* na = new Articulation(articSym, cr->score());
-      if (dir == "up") {
-            na->setUp(true);
-            na->setAnchor(ArticulationAnchor::TOP_STAFF);
-            na->setPropertyFlags(Pid::ARTICULATION_ANCHOR, PropertyFlags::UNSTYLED);
+
+      if (!dir.isNull()) {
+            na->setUp(dir.isEmpty() || dir == "up");
+            if (dir.isEmpty()) // Set as an implicit value
+                  na->setPropertyFlags(Pid::DIRECTION, PropertyFlags::STYLED);
+            else // Set as an explicit value
+                  na->setPropertyFlags(Pid::DIRECTION, PropertyFlags::UNSTYLED);
             }
-      else if (dir == "down") {
-            na->setUp(false);
+
+      if (place == "above" || dir.isEmpty() || dir == "up")
+            na->setAnchor(ArticulationAnchor::TOP_STAFF);
+      else if (place == "below" || dir == "down")
             na->setAnchor(ArticulationAnchor::BOTTOM_STAFF);
+      if (dir.isEmpty()) // Set as an implicit value
+            na->setPropertyFlags(Pid::ARTICULATION_ANCHOR, PropertyFlags::STYLED);
+      else if (!place.isNull() || !dir.isNull()) // Set as an explicit value
             na->setPropertyFlags(Pid::ARTICULATION_ANCHOR, PropertyFlags::UNSTYLED);
-            }
-      cr->add(na);
-      }
 
-//---------------------------------------------------------
-//   addAllArticulationsToChord
-//---------------------------------------------------------
-
-static void addAllArticulationsToChord(ChordRest* cr, std::vector<SymId> symbols)
-      {
-      for (const auto symbol : symbols) {
-            addArticulationToChord(cr, symbol, "");
-            }
+      cr->add(na);
       }
 
 //---------------------------------------------------------
 //   addFermataToChord
 //---------------------------------------------------------
 
-static void addFermataToChord(ChordRest* cr, SymId articSym, bool up)
+/**
+ Add a MusicXML fermata.
+ Note: MusicXML common.mod: "The fermata type is upright if not specified."
+ */
+
+static void addFermataToChord(ChordRest* cr, SymId articSym, QString direction)
       {
       Fermata* na = new Fermata(articSym, cr->score());
       na->setTrack(cr->track());
-      na->setPlacement(up ? Placement::ABOVE : Placement::BELOW);
+      if (!direction.isNull()) {
```

Comment:
QString::isNull() is depricated. Remove or replace by isEmpty() or comparison with "" where required. Note isMull found in many places, remove all.

Reply:
even though `isNull` is not recommended, i need to make distinction between `QString()` (when XML attribute is not present) and `QString("")` (when attribute has explicit empty value). i have removed it in the places where it was used both isNull and isEmpty

Reply:
'if (!direction)' might help here

Reply:
`if (!direction)` might help here?
Maybe better place a comment here as to why you need `isNull()` here.

Reply:
As a sanity check question, do empty attributes actually occur in MusicXML files and do they have a different meaning from a missing attribute ? If not, isNull is not required. If the distinction is necessary (which I am not aware of, but I may be wrong), you could also solve it by passing the annotation as const& to addFermataToChord() and check for presence of the type attribute.